runtime/manta/src/lib.rs

```diff
@@ -331,6 +336,85 @@ impl pallet_timestamp::Config for Runtime {
     type WeightInfo = weights::pallet_timestamp::SubstrateWeight<Runtime>;
 }
 
+/// Only callable after `set_validation_data` is called which forms this proof the same way
+fn relay_chain_state_proof() -> cumulus_pallet_parachain_system::RelayChainStateProof {
+    let relay_storage_root = ParachainSystem::validation_data()
+        .expect("set in `set_validation_data`")
+        .relay_parent_storage_root;
+    let relay_chain_state =
+        ParachainSystem::relay_state_proof().expect("set in `set_validation_data`");
+    cumulus_pallet_parachain_system::RelayChainStateProof::new(
+        ParachainInfo::get(),
+        relay_storage_root,
+        relay_chain_state,
+    )
+    .expect("Invalid relay chain state proof, already constructed in `set_validation_data`")
+}
+pub struct BabeDataGetter;
+impl pallet_randomness::GetBabeData<u64, Option<Hash>> for BabeDataGetter {
+    // Tolerate panic here because only ever called in inherent (so can be omitted)
+    fn get_epoch_index() -> u64 {
+        if cfg!(feature = "runtime-benchmarks") {
+            // storage reads as per actual reads
+            let _relay_storage_root = ParachainSystem::validation_data();
+            let _relay_chain_state = ParachainSystem::relay_state_proof();
+            const BENCHMARKING_NEW_EPOCH: u64 = 10u64;
+            return BENCHMARKING_NEW_EPOCH;
+        }
+        relay_chain_state_proof()
+            .read_optional_entry(cumulus_primitives_core::relay_chain::well_known_keys::EPOCH_INDEX)
+            .ok()
+            .flatten()
+            .expect("expected to be able to read epoch index from relay chain state proof")
+    }
+    fn get_epoch_randomness() -> Option<Hash> {
+        if cfg!(feature = "runtime-benchmarks") {
+            // storage reads as per actual reads
+            let _relay_storage_root = ParachainSystem::validation_data();
+            let _relay_chain_state = ParachainSystem::relay_state_proof();
+            let benchmarking_babe_output = Hash::default();
+            return Some(benchmarking_babe_output);
+        }
+        relay_chain_state_proof()
+            .read_optional_entry(
+                // We use randomness from one relaychain epoch ago (4hrs Polkadot, 1hr Kusama) in combination with a (longer) `DrawingFreezeout` in pallet lottery
+                // to prevent manuipulation of the winning set by participants of the lottery
+                // NOTE: We operate under the following assumption https://github.com/Manta-Network/Manta/blob/f7b1a74e0261cc6a73081a9a86f8cd87aef8ea27/pallets/randomness/README.md#L85,
```

Comment:
`The only incentive to prevent the validator from skipping the block is the block rewards. So the randomness value is only economically safe if the value at stake is lower than a block reward.` Doesn't that mean that if the lottery is higher than a block reward, then the randomess is not good ?

Reply:
it means the last validator of the epoch before the drawing gets two tickets instead of one because he can choose to skip his block to re-roll.
a) This is fairly low-impact
b) i've been told not to worry about fairness of the randomness algorithm itself

Reply:
ahaa ok i see, was thinking something else